File: app/playRepository/SvnCommit.java
/**
 * Yobi, Project Hosting SW
 *
 * Copyright 2012 NAVER Corp.
 * http://yobi.io
 *
 * @author Yi EungJun
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package playRepository;

import models.User;

import java.util.Date;
import java.util.TimeZone;

import org.tmatesoft.svn.core.SVNLogEntry;

public class SvnCommit extends Commit {
    SVNLogEntry entry;

    public SvnCommit(SVNLogEntry entry) {
        this.entry = entry;
    }

    private long getIdAsLong() {
        return entry.getRevision();
    }

    @Override
    public String getMessage() {
        return entry.getMessage();
    }

    @Override
    public User getAuthor() {
        return User.findByLoginId(getAuthorName());
    }

    @Override
    public String getAuthorName() {
        return entry.getAuthor();
    }

    @Override
    public String getId() {
        return String.valueOf(getIdAsLong());
    }

    @Override
    public String getShortMessage() {
        String message = getMessage();
        if (message != null && !message.isEmpty()) {
            String[] lines = message.trim().split("\n");
            return lines.length > 0 ? lines[0] : "";
        } else {
            return "";
        }
    }

    @Override
    public String getAuthorEmail() {
        return null;
    }

    @Override
    public Date getAuthorDate() {
        return entry.getDate();
    }

    @Override
    public TimeZone getAuthorTimezone() {
        return null;
    }

    @Override
    public String getCommitterName() {
        return getAuthorName();
    }

    @Override
    public String getCommitterEmail() {
        return null;
    }

    @Override
    public Date getCommitterDate() {
        return getAuthorDate();
    }

    @Override
    public TimeZone getCommitterTimezone() {
        return null;
    }

    @Override
    public int getParentCount() {
        if (getIdAsLong() > 1) {
            return 1;
        } else {
            return 0;
        }
    }

    @Override
    public String getShortId() {
        return getId();
    }
}

File: src/main/kotlin/com/github/yonaprojects/yona/domain/vcs/SvnCommit.kt
package com.github.yonaprojects.yona.domain.vcs

import com.github.yonaprojects.yona.domain.gpgkey.GpgVerificationStatus
import com.github.yonaprojects.yona.domain.user.User
import org.tmatesoft.svn.core.SVNLogEntry
import java.util.Date
import java.util.TimeZone

class SvnCommit(
    private val entry: SVNLogEntry,
    private val userResolver: (String) -> User?
) : Commit() {

    private val idAsLong: Long
        get() = entry.revision

    override fun getMessage(): String? {
        return entry.message
    }

    override fun getAuthor(): User? {
        val authorName = getAuthorName() ?: return null
        return userResolver(authorName)
    }

    override fun getAuthorName(): String? {
        return entry.author
    }

    override fun getId(): String {
        return idAsLong.toString()
    }

    override fun getShortMessage(): String {
        val msg = getMessage()
        return if (!msg.isNullOrEmpty()) {
            val lines = msg.trim().split("\n")
            if (lines.isNotEmpty()) lines[0] else ""
        } else {
            ""
        }
    }

    override fun getAuthorEmail(): String? {
        return null
    }

    override fun getAuthorDate(): Date? {
        return entry.date
    }

    override fun getAuthorTimezone(): TimeZone? {
        return null
    }

    override fun getCommitterName(): String? {
        return getAuthorName()
    }

    override fun getCommitterEmail(): String? {
        return null
    }

    override fun getCommitterDate(): Date? {
        return getAuthorDate()
    }

    override fun getCommitterTimezone(): TimeZone? {
        return null
    }

    override fun getParentCount(): Int {
        return if (idAsLong > 1) 1 else 0
    }

    override fun getShortId(): String {
        return getId()
    }

    override fun getGpgVerificationStatus(): GpgVerificationStatus {
        // SVN은 이 앱에서 GPG 커밋 서명 개념이 없다(계획 문서 범위 밖).
        return GpgVerificationStatus.UNSIGNED
    }
}
